src/noflo-component-access.js

```diff
@@ -0,0 +1,143 @@
+ * also present for arbitrary assignments and is shared for all facades of the
+ * same Component instance.
+ * This facade mimic the main usage of http://noflojs.org/api/Component.html
+ * Usage:
+ *  var component = componentAccess(new noflo.Component());
+ */
+function facadeComponent(component) {
+    return _.extend({
+        get inPorts() {
+            return _.mapObject(_.pick(component.inPorts, isInPort), facadePort);
+        },
+        get outPorts() {
+            return _.mapObject(_.pick(component.outPorts, isOutPort), facadePort);
+        },
+        get rpf() {
+            return component.rpf = component.rpf || {};
+        }
+    }, facadeFunctions(component,
+        'getDescription',
+        'getIcon'
+    ), facadeEventEmitter(component));
+}
+
+/**
+ * Create a noflo Port facade from a given noflo Port that includes only a limited
+ * subset of properties and functions.
+ * This facade mimic the main usage of http://noflojs.org/api/OutPort.html
+ */
+function facadePort(port) {
+    return _.extend({
+        // port.nodeInstance might not be assigned yet
+        get nodeInstance() {
+            return port.nodeInstance && facadeComponent(port.nodeInstance)
+        },
+        get name() {
+            return port.name;
+        }
+    }, facadeFunctions(port,
+        'getDescription',
+        'getDataType',
+        'hasDefault',
+        'isAddressable',
+        'isBuffered',
+        'isRequired',
+        'isCaching',
+        'isConnected',
+        'connect',
+        'beginGroup',
+        'send',
+        'endGroup',
+        'disconnect',
+        'checkRequired'
+    ), facadeEventEmitter(port));
+}
+
+/**
+ * Returns a facade of the given EventEmitter. The result is not an EventEmitter,
+ * but includes the same functions. Each function call is passed to the underlying
+ * EventEmitter, while any provided functions are bound to the original context
+ * when they are called. This gives the behaviour of an EventEmitter without
+ * actually creating one with a distinct state.
+ * @see https://nodejs.org/dist/latest-v4.x/docs/api/events.html
+ */
+function facadeEventEmitter(emitter) {
+    return facadeFunctions(emitter,
+        'addListener',
+        'emit',
+        'getMaxListeners',
+        'listenerCount',
+        'listeners',
+        'on',
+        'once',
+        'removeAllListeners',
+        'removeListener',
+        'setMaxListeners'
+    );
+}
+
+/**
+ * Picks all the given functions from object and returns the set bounded to the
+ * original object context. If any argument is passed to those functions that is
+ * also a function it is bound to the called context.
+ * @param object with functions, i.e. noflo.Component, noflo.InPort/OutPort
+ * @param keys explicit names of functions to include in result
+ */
+function facadeFunctions(object /* keys */) {
+    return _.mapObject(_.pick(object, arguments), function(fn) {
```

Comment:
Should that be the following, to avoid trying to pick 'object' as a key?
<pre>
return _.mapObject(_.pick(object, _.toArray(arguments).slice(1)), function(fn) {
</pre>